Adding functionalities for cleaning up!!

senthil10 opened the pull request:
Currently cleaning up of productions `archive`(runs directory), `illumina`(projects directory) and `project`(analysis directory) are not cleaned up automatically or not clean very obvious if they are being cleaned. So this pull request has functionalities/methods which are intended to run once a month to main the disk quota under control. 

@vezzi and @guillermo-carrasco : have tried it, even in uppmax (of course with dry run option). Looks like it does what it supposed to do. 

But definitely need suggestions for improvements from all programming gurus :)

guillermo-carrasco commented on taca/storage/cli.py:
```diff
@@ -16,19 +15,30 @@ def storage(ctx, days, run):
 
 
 @storage.command()
+@click.option('--site', type=click.Choice(['swestore','archive','illumina','analysis','nas','proc']),
```
short option as well? `-s`

guillermo-carrasco replied:
and I'd rather prefer to be specific: `proc`--> `processing` or even `processing-server`

senthil10 replied:
Even I was concerned about that, was concerned about your preference as you would implement it. But now I know so i will change it :smile: :+1:

guillermo-carrasco commented on taca/storage/storage.py:
```diff
@@ -2,16 +2,27 @@
+
+def cleanup_nas(days):
+    """
+        Will move the finished runs in NASes and processing server
```
Indentation is a bit weird here, check [this](https://www.python.org/dev/peps/pep-0008/#documentation-strings)

guillermo-carrasco commented on taca/storage/storage.py:
```diff
@@ -28,6 +39,12 @@ def cleanup(days):
+    """
+        Send runs (as archives) in NAS nosync to swestore for backup
+
+        :param int days: number fo days to check threshold
```
Good docstrings ^^, tack!

guillermo-carrasco commented on taca/storage/storage.py:
```diff
@@ -64,6 +81,93 @@ def archive_to_swestore(days, run=None):
+            if dry_run:
+                LOG.info('Will remove file {} from swestore'.format(run))
+                continue
+            misc.call_external_command('irm -f {}'.format(run))
```
This is a question: this works in the NASs because the iCommands are part of the standard path (i.e no module needs to be loaded), I guess that in Uppmax is a **requirement** to have loaded `module load irods/swestore_legacy`, is it?

senthil10 replied:
@guillermo-carrasco : Yes it have to be loaded before, but in case if the script is called without loading `module load irods/swestore_legacy` it would throw an `OSError` and fails when executing `runs = filesystem.list_runs_in_swestore(path=config.get('cleanup').get('swestore'))`

guillermo-carrasco replied:
:+1:

mariogiov replied:
You could also use [this slick function](https://github.com/NationalGenomicsInfrastructure/ngi_pipeline/blob/devel/ngi_pipeline/utils/filesystem.py#L26) that @ewels wrote to load modules from within Python

senthil10 replied:
Tack @mariogiov , seems interesting will look into it :)

guillermo-carrasco commented on taca/storage/storage.py:
```diff
@@ -64,6 +81,93 @@ def archive_to_swestore(days, run=None):
+    LOG = get_logger()
+    root_dir = config.get('cleanup').get(site)
+    deleted_log = config.get('cleanup').get('deleted_log')
+    assert os.path.exists(os.path.join(root_dir,deleted_log)), "Log directory {} doesn't exist in {}".format(deleted_log,root_dir)
```
nice one :-)

guillermo-carrasco commented on taca/storage/storage.py:
```diff
@@ -64,6 +81,93 @@ def archive_to_swestore(days, run=None):
+    deleted_log = config.get('cleanup').get('deleted_log')
+    assert os.path.exists(os.path.join(root_dir,deleted_log)), "Log directory {} doesn't exist in {}".format(deleted_log,root_dir)
+    log_file = os.path.join(root_dir,"{fl}/{fl}.log".format(fl=deleted_log))
+    PRO_RE = '[a-zA-Z]+\.[a-zA-Z]+_\d{2}_\d{2}'
```
`PROJECT_RE` there is no need to shorten this and is more clear ;-) that's only my opinion though

vezzi replied:
Maybe more appropriate here
taca/utils/filesystem.py
close to RUN_RE

senthil10 replied:
@vezzi : I thought about that too, and said to myself if someone else think alike will move it there :wink:

ewels replied:
@guillermo-carrasco personally I prefer the variable name
```
SCIENCE_FOR_LIFE_LABORATORY_NATIONAL_GENOMICS_INFRASTRUCTURE_PROJECT_NAME_REGULAR_EXPRESSION_PATTERN
```
:smiling_imp: (but yes, I agree)

guillermo-carrasco replied:
that's all you can do @ewels ? what a disappointment... 

`SCIENCE_FOR_LIFE_LABORATORY_STOCKHOLM_NODE_NATIONAL_GENOMICS_INFRASTRUCTURE_PROJECT_NAME_REGULAR_EXPRESSION_PATTERN` and I can keep one...

P.S: I second to move the pattern to `filesystem.py`

vezzi commented on taca/storage/cli.py:
```diff
@@ -16,19 +15,30 @@ def storage(ctx, days, run):
+    days = params.get('days')
+    if not days:
+        days = st.site_check_days[site]
+    if site == 'nas':
```
there is no option for `proc`

senthil10 replied:
@vezzi : Yes, me and @guillermo-carrasco discussed about that. He will implement it as he know better and have already done that.

vezzi replied:
ok, anyway can you mock it (simply write on LOG that this is not yet supported)

guillermo-carrasco replied:
or raise a `NotImplementedError`

guillermo-carrasco commented on taca/storage/storage.py:
```diff
@@ -64,6 +81,93 @@ def archive_to_swestore(days, run=None):
+    db_config = config.get('statusdb',{})
+
+    # make a connection for project db #
+    pcon = statusdb.ProjectSummaryConnection(**db_config)
```
ahhh noo why did you implemented this call like this? I don't like when parameters come in the form of `**params`, its just so confusing

senthil10 replied:
Cool!! Do you prefer specifying each param individually, and that is also another thing I wanted to discuss.

we can also call it like `pcon = statusdb.ProjectSummaryConnection()` if we have a statusdb config file here `~/.ngi_config/statusdb`

which one is better ?

ewels replied:
+1 for disliking `**vars` syntax. Can be tricky to figure out where stuff comes from.

guillermo-carrasco replied:
If the `~/.ngi_config/statusdb` is required in `statusdb` package, go for the second one. Otherwise the first one, specifying all parameters I think is still better than `**db_config` :-)

senthil10 replied:
Yup, config file for `statusdb` is mandatory (thanks to @ewels ). So I will go for the second :+1:

ewels replied:
I think that this is cleaner - the point of the statusdb package is to centralise all database interactions, so it seems strange to abstract the connection config away from this.

guillermo-carrasco replied:
awesome :+1:

guillermo-carrasco commented on taca/storage/storage.py:
```diff
@@ -64,6 +81,93 @@ def archive_to_swestore(days, run=None):
+        with filesystem.chdir(root_dir):
+            for run in runs:
+                fcid = run.split('_')[3]
+                fcid = fcid if re.search('-',fcid) else fcid[1:]
```
is this a check for MiSeqs?

guillermo-carrasco commented on taca/storage/storage.py:
```diff
@@ -64,6 +81,93 @@ def archive_to_swestore(days, run=None):
+                try:
+                    ssheet = [c for c in os.listdir(run) if re.search('{}.csv|SampleSheet.csv'.format(fcid),c)][0]
+                except IndexError:
+                    LOG.warn("Could not find expected samplesheet for run {}, so SKIPPING..")
```
you're missing a format here

senthil10 replied:
hahaha,. I do,. :joy:

guillermo-carrasco commented on taca/storage/storage.py:
```diff
@@ -64,6 +81,93 @@ def archive_to_swestore(days, run=None):
+        try:
+            shutil.rmtree(os.path.join(root_dir,item))
+            LOG.info('Removed project {} from {}'.format(item,root_dir))
+            mode = 'a' if os.path.exists(log_file) else 'w'
```
you can just open a file with `wa`mode and it will create it if it doesn't exist. No need for this conditional :-)

senthil10 replied:
I tried that first, but for some reason it dint work :( 
It just kept overwriting the file. But will give it a try again and see more carefully :+1:

guillermo-carrasco commented on taca/storage/storage.py:
```diff
@@ -114,3 +218,31 @@ def _send_to_swestore(f, dest, remove=True):
     os.remove("{}.archiving".format(run.split('.')[0]))
+
+
+def get_closed_projects(projs,pj_con,days,logger=None):
```
don't quite understand why are you passing logger as a parameter here :-/

senthil10 replied:
I wanted to LOG some stuff if necessary, but since `LOG` is initiated within different methods, I was not sure if it is accessible in this method. I mean would the following example work ?

```
def do_something(params):
    LOG = logger()
    do_another_thing(whatever)


def do_another_thing(params):
    print params
    LOG.info('will this work ?')
```

Will this execute without error or would it fail saying `LOG` global name not defined ? I always have this doubt if a `name` is defined in parent method where another method is called, and can the `name` be accessed within the child method without having to pass ?

guillermo-carrasco replied:
no that wouldn't work, you would have to execute `LOG = get_logger()` in `do_another_thing()` too, this is called "scope of variables" there are lots of docs about this out there if you're bored ;-)

senthil10 replied:
@guillermo-carrasco : Thats my reason why I passed `logger` as argument :) But do you prefer to initiate logger within the new method instead of passing it as argument ??

guillermo-carrasco replied:
yeah I think so, its more consistent with the rest of the code. 

On the other hand... I may think on a decorator for both LOG and config, I'm constantly loading them.

senthil10 replied:
Yup, makes sense to have them. As currently all the methods will need them, good to have them globally.

guillermo-carrasco commented on taca/utils/misc.py:
```diff
@@ -66,3 +65,18 @@ def call_external_command_detached(cl, with_log_files=False):
+        return None
+    t = datetime.today()
+    time_dif = t - d
+    return(time_dif.days)
```
no parenthesis are needed

senthil10 replied:
So used to `R` :sweat_smile:

ewels replied:
Maybe not, but they do _look_ nice. If you could added a semi-colon it'd be even better..

guillermo-carrasco replied:
they look nice?? stop trolling @ewels , haha

ewels replied:
:trollface:

guillermo-carrasco commented:
Good work @senthil10 !

I'm done with the review, just a couple of styling comments to add to the inline ones:

* After a comma always follows a space, so for example def `list_runs_in_swestore(path,pattern=RUN_RE,no_ext=False):` would be better as `def list_runs_in_swestore(path, pattern=RUN_RE, no_ext=False):`
* There should'n be any spaces on the opening and closing bracket in list comprehensions, so for example `[ a for a in some_iterable ]` would be better as `[for a in some_iterable]`

Finally, something that I am not sure about (but I don't have a clear suggestion) is the global check_old_days that you've declared in `storage.py`. Maybe would be better in a config file?

Anyway, bra jobbat!

vezzi commented:
@senthil10 looks good only few things to fix.
two things to think about but maybe for the next version:
 * preproc cleaning (b5, preproc1, and preproc2) seems to not been implemented. The cleaning on preproc serves is in some way conditioned by the fact that the run is transferred to Uppmax and archived on swestore, and bcl2fastq has been correctly run.... in general I would say that if the demultiplexed files are one week older it means that they have been correctly demultiplexed.... but this is a rule of thumb
 * there will be soon the need to implement this on Nestor and Xten, therefore SampleSheet processing will change as SampleSheet is different, and the demultiplexing folder structure is different as well..... just to let you know

If today you can fix @guillermo-carrasco notes we can deploy it right away

Great job
FV

ewels commented on taca/utils/misc.py:
```diff
@@ -66,3 +65,18 @@ def call_external_command_detached(cl, with_log_files=False):
+    except ValueError:
+        return None
+    t = datetime.today()
+    time_dif = t - d
```
`time_dif` -> `time_diff`
`</pedant>`

Also why bother with all of this variable assignment? Seems like you could simplify it down quite a bit and it would be clearer:
```python
try:
   return datetime.today() - datetime.strptime(date, date_format)
except ValueError:
   return None
```

senthil10 replied:
hehe, seems like it would work as well :joy:  tack @ewels

senthil10 commented:
Thank you guys for the wonderful/insightful review, I will make a commit reflecting on all your suggestions (except the one made by @ewels on the variable name :P) 

@guillermo-carrasco : Regarding the `site_check_days` i.e. default check days dictionary. I want that to be discussed. I also had another idea (like you suggested) having git in config file. Will look something like

```
cleanup:
    archive:
        root: /path/to/root
        check_days: 90
```
Would that be nice ?

@vezzi : Yes, I will Implement a WIP method for `processing server` :+1:

guillermo-carrasco commented:
@senthil10 yes, in the config file would be nicer :-)

Just one more thing: Remember to include statusdb in `requirements.txt`, and it would be so awesome if you could upload `statusdb` to Pypi to be able to resolve the dependency automatically with `pip install statusdb`, [but don't stress with that](https://docs.python.org/2/distutils/packageindex.html)

senthil10 commented:
@vezzi  and @guillermo-carrasco : I have fixed all the comments and also added some new stuff extra. Made couple of commits, but [this one] (https://github.com/senthil10/TACA/commit/43684a07646082bf043bb8eaa52d397626af160a) is main with all major changes. I have also made some commits after with one line fixes.

Anyways major thing I added is `query_yes_no` from `scilifelab.utils`, as the script is removing stuff I am very nervous about that. So I added a prompt, so it will ask the user for confirmation if `-d` option is called with days that are less than default threshold. Hope that is fine(?). I was just worried about the typo, what if someone (probably me) run cleanup with `-d 8` actually which was supposed to be `-d 80`.

@vezzi : Also for runs in `archive` directory, I am checking if they are in swestore as well (as double check). If you think we don't need this I can take away :)

senthil10 commented:
P.S: Travis fails cause I added `statusdb` to `requirements`. Will look into @guillermo-carrasco suggestion to add `statusdb` to `Pypi`. But any quick fix ? :sweat_smile:

ewels commented:
@senthil10 @guillermo-carrasco I think we can add the statusdb repository to the install file can't we, instead of adding it to PyPi? This seems like a bit of a niche tool to be adding to a public tools database..

https://pip.pypa.io/en/latest/reference/pip_install.html#vcs-support

So I think the following should work for `requirements.txt`?
```
git+https://github.com/SciLifeLab/statusdb.git#egg=statusdb
```

senthil10 commented:
@ewels :  I too agree with you. But thanks for the suggestion. Made a commit with that change, hope it will fix travis build. 

@guillermo-carrasco : Isn't it too small/focused to put it on `Pypi` ? :)

ewels commented:
Apparently although pip installs that kind of url, [setuptools doesn't](http://stackoverflow.com/questions/14737500/get-pip-to-work-with-git-and-github-repository)

Try this instead:

```
https://github.com/SciLifeLab/statusdb/archive/master.zip#egg=statusdb
```

senthil10 commented:
I dont know When I try this 

` pip install git+https://github.com/SciLifeLab/statusdb.git#egg=statusdb`

it works, but when I try it using `setup_tools` it complains (even without `git+`):angry:

ewels commented:
Yup, see my comment above :)

> although pip installs that kind of url, setuptools doesn't

The one I pasted _should_ work, though I haven't tried it.

pip knows how to handle `.git` files, but setuptools doesn't. If you use a URL with a `.zip` extension it should be ok though.

guillermo-carrasco commented:
I don't see the problem on putting it on pypi to be honest, is not that we have to pay for it or something, and is just a couple of commands do, TACA is also quite SciLife specific and its on pypi ;-)

ewels commented:
Fair enough. Seems easy to add the github url to the install scripts though as well? And no-one outside of NGI is ever going to use these packages, right?

guillermo-carrasco commented:
But the problem is that you can't add the github url to the `requirements.txt` file because it won't work. I mean, you can do `pip install git+https://github.com/SciLifeLab/statusdb.git#egg=statusdb`, but this doesn't work with setuptools for the dependency resolution. 

I really don't care how it is done, I just want to not have to worry about having to install several repos to have this one working. However it is done, either `python setup.py install` or `pip install taca` should install a functional tool.

ewels commented:
Am I getting this wrong here or are you guys just completely missing my point? :)

>> although pip installs that kind of url, setuptools doesn't

>The one I pasted _should_ work, though I haven't tried it.

The second URL I pasted wasn't the GitHub `.git` url that works with `pip`, it's the `.zip` file download link which should work with setuptools.

This works for pip:
```
pip install git+https://github.com/SciLifeLab/statusdb.git#egg=statusdb
```

This should work in requirements.txt:
```
https://github.com/SciLifeLab/statusdb/archive/master.zip#egg=statusdb
```

Although they both start with `https://github.com`, they are not the same link...

..yeah?

senthil10 commented:
Hej,

@ewels : Yes, I see your point I even made a [commit](https://github.com/senthil10/TACA/commit/7886654ec1db82f01d2f27f43c971003f1ff14d9). But the real problem is we cant give links in `requirement.txt` or to `install_requires` while calling `setup()` in `setup.py`. Thats the reason why `Travis` fails, I get the same error if try to install `TACA` with that link in requirement.txt.

So from [here](http://stackoverflow.com/questions/19738085/why-isnt-setup-py-dependency-links-doing-anything), seems like we have to give the links to `dependency_links` in `setup()` and also mention the package name in `install_requires`. So for us it would be something like 

```
dependency_links = ['https://github.com/SciLifeLab/statusdb/archive/master.zip#egg=statusdb'],
install_requires = ['statusdb']
```

I tried this and it seems to work and have implemented it in this [commit](https://github.com/senthil10/TACA/commit/1131b0d39c622509c311062f6bb957aa7c76a527). Even [Travis build](https://travis-ci.org/SciLifeLab/TACA/builds/54438286) went through.Thank you for your pointers :)

@guillermo-carrasco :  Let me know what do you think :)

ewels commented:
Super nice, good work @senthil10! :grinning:

guillermo-carrasco commented:
Good job @senthil10 ! Still don't understand this reluctance to pypi but well, won't force you. 

@vezzi you can merge if you've finished your review :-)

senthil10 commented:
@guillermo-carrasco : I dont mind putting `statusdb` in `Pypi` (like you said I just little hesitant). But why I fall back to this is, we need `taca storage cleanup` in production soon and this method was faster for me :) I will look into that `Pypi` thingy as well :+1: 

@vezzi and @ewels : Tack sa mycket :)
